Forge.TreeWalker/src/SubroutineAction.cs

```diff
@@ -0,0 +1,113 @@
+    ///   3. Explicitly calls UpdateKeyPrefix on the ForgeState with the updated Subroutine SessionId.
+    ///   4. Walks the tree starting at the CurrentTreeNode if persisted, otherwise RootTreeNodeKey if given, otherwise "Root" as default.
+    ///   5. Returns the Subroutine's last ActionResponse if available, otherwise returns the tree walker session Status.
+    /// </summary>
+    [ForgeAction(InputType: typeof(SubroutineInput))]
+    public class SubroutineAction : BaseAction
+    {
+        /// <summary>
+        /// The tree walker parameters of the parent tree walker session.
+        /// Used to call InitializeSubroutineTree to get the initialized TreeWalkerSession for this Subroutine.
+        /// </summary>
+        private TreeWalkerParameters parameters { get; set; }
+
+        /// <summary>
+        /// Instantiates a SubroutineAction with the required parameters.
+        /// </summary>
+        /// <param name="parameters">The tree walker parameters of the parent tree walker session.</param>
+        public SubroutineAction(TreeWalkerParameters parameters)
+        {
+            this.parameters = parameters;
+        }
+
+        /// <summary>
+        /// The RunAction method is called when Forge encounters an ActionName while walking the tree.
+        /// </summary>
+        /// <param name="actionContext">The action context holding relevant information for this Action.</param>
+        public override async Task<ActionResponse> RunAction(ActionContext actionContext)
+        {
+            SubroutineInput input = (SubroutineInput)actionContext.ActionInput;
+
+            // Rehydrate the subroutine's SessionId if previously persisted.
+            SubroutineIntermediates intermediates = await actionContext.GetIntermediates<SubroutineIntermediates>();
+            if (intermediates == null)
+            {
+                intermediates = new SubroutineIntermediates()
+                {
+                    SessionId = Guid.NewGuid()
+                };
+
+                await actionContext.CommitIntermediates<SubroutineIntermediates>(intermediates);
+            }
+
+            // Initialize TreeWalkerSession for this subroutine.
+            TreeWalkerSession subroutineSession = this.parameters.InitializeSubroutineTree(input, intermediates.SessionId, this.parameters);
+
+            // Update KeyPrefix of ForgeState for state persistence separation.
+            subroutineSession.Parameters.ForgeState.UpdateKeyPrefix(subroutineSession.Parameters.RootSessionId, subroutineSession.Parameters.SessionId);
+
+            // Walk tree starting at CurrentTreeNode if persisted, otherwise RootTreeNodeKey if given, otherwise "Root" as default.
+            string currentTreeNode = await subroutineSession.GetCurrentTreeNode() ?? subroutineSession.Schema.RootTreeNodeKey;
+
+            // WalkTree may throw exceptions. We let this happen to allow for possible retry handling.
```

Comment:
"possible retry handling" will the code of retry reside inside the forge tree walker or on the anvil side?
also what are cases where it will not retry?

Reply:
Here, the retry handling refers to the native RetryPolicy in Forge for all Actions. You can read more about it here: https://github.com/microsoft/Forge/wiki/ForgeAction-Authoring---How-To-Guide#Forge-Behavior-and-how-it-changes-based-on-RunAction-response-and-ForgeSchema-configuration